tests/transforms/test_tape_expand.py

```diff
@@ -328,3 +329,229 @@ class NonDiffPhaseShift(qml.PhaseShift):
+        decomp_dev = qml.device("default.qubit", wires=2, custom_decomps=custom_decomps)
+        decomp_qnode = qml.QNode(circuit, decomp_dev, expansion_strategy="device")
+        _ = decomp_qnode()
+        decomp_ops = decomp_qnode.qtape.operations
+
+        assert len(decomp_ops) == 7
+
+        assert decomp_ops[0].name == "RX"
+        assert decomp_ops[0].parameters[0] == 0.1
+        assert decomp_ops[0].wires == Wires(0)
+
+        assert decomp_ops[1].name == "RX"
+        assert decomp_ops[1].parameters[0] == 0.2
+        assert decomp_ops[1].wires == Wires(1)
+
+        assert decomp_ops[2].name == "RZ"
+        assert np.isclose(decomp_ops[2].parameters[0], np.pi)
+        assert decomp_ops[2].wires == Wires(1)
+
+        assert decomp_ops[3].name == "RY"
+        assert np.isclose(decomp_ops[3].parameters[0], np.pi / 2)
+        assert decomp_ops[3].wires == Wires(1)
+
+        assert decomp_ops[4].name == "CZ"
+        assert decomp_ops[4].wires == Wires([0, 1])
+
+        assert decomp_ops[5].name == "RZ"
+        assert np.isclose(decomp_ops[5].parameters[0], np.pi)
+        assert decomp_ops[5].wires == Wires(1)
+
+        assert decomp_ops[6].name == "RY"
+        assert np.isclose(decomp_ops[6].parameters[0], np.pi / 2)
+        assert decomp_ops[6].wires == Wires(1)
+
+    def test_custom_decomp_with_adjoint(self):
+        """Test that applying an adjoint in the circuit results in the adjoint
+        undergoing the custom decomposition."""
+
+        def circuit():
+            # Adjoint is RX(-0.2), so expect RY(-0.2) H
+            qml.adjoint(qml.RX)(0.2, wires="a")
+            return qml.expval(qml.PauliZ("a"))
+
+        custom_decomps = {qml.RX: custom_rx}
+        decomp_dev = qml.device("default.qubit", wires="a", custom_decomps=custom_decomps)
+        decomp_qnode = qml.QNode(circuit, decomp_dev, expansion_strategy="device")
+        _ = decomp_qnode()
+        decomp_ops = decomp_qnode.qtape.operations
+
+        assert len(decomp_ops) == 2
+
+        assert decomp_ops[0].name == "RY"
+        assert decomp_ops[0].parameters[0] == -0.2
+        assert decomp_ops[0].wires == Wires("a")
+
+        assert decomp_ops[1].name == "Hadamard"
+        assert decomp_ops[1].wires == Wires("a")
+
+    def test_custom_decomp_with_control(self):
+        """Test that applying a controlled version of a gate results in the
+        controlled version of a decomposition."""
+
+        def circuit():
+            qml.ctrl(qml.Hadamard, control=0)(wires=1)
+            return qml.expval(qml.PauliZ(0))
+
+        custom_decomps = {qml.Hadamard: custom_hadamard}
+        decomp_dev = qml.device("default.qubit", wires=2, custom_decomps=custom_decomps)
+        decomp_qnode = qml.QNode(circuit, decomp_dev, expansion_strategy="device")
+        _ = decomp_qnode()
+        decomp_ops = decomp_qnode.qtape.operations
+
+        assert len(decomp_ops) == 2
+
+        assert decomp_ops[0].name == "CRZ"
```

Comment:
Honestly I was kind of surprised this worked out-of-the-box. It seems like the operation being controlled gets decomposed first, and then each gate in the decomposition gets the control applied.

Reply:
@glassnotes I'm trying to think of any other tests cases, but it sounds like you've covered pretty much all! Only other thing I can think of is a decomposition where:

- template A is decomposed into template B
- gate C is decomposed into gate D

and then decomposing a circuit that has template A?

Reply:
I also have to add gradient tests, which I just totally forgot about :sweat_smile:

Reply:
sure! Although my gut feeling is that this is one of the parts of the codebase where gradient tests aren't needed, since this PR doesn't introduce any classical processing.